src/Native/System.Native/pal_networking.cpp

```diff
@@ -2049,6 +2049,27 @@ SystemNative_SetSockOpt(int32_t socket, int32_t socketOptionLevel, int32_t socke
         return PAL_ENOTSUP;
     }
 
+#ifndef WIN32
+    if ((optLevel == IPPROTO_IP) && (optName == IP_MULTICAST_IF) && (optionLen == sizeof(in_addr)))
+    {
+        /* if the value of the IP_MULTICAST_IF is an address in the 0.x.x.x block
+           the value is interpreted as an interface index */
+        if (optionValue[0] == 0)
+        {
+#ifdef HAVE_IP_MREQN
+            struct ip_mreqn mreq =
+                   {.imr_multiaddr = {.s_addr = 0},
+                    .imr_address = {.s_addr = 0},
+                    .imr_ifindex = (optionValue[1] << 16) + (optionValue[2] << 8) + optionValue[3]};
+            int err = setsockopt(socket, optLevel, optName, &mreq, static_cast<socklen_t>(sizeof(mreq)));
+            return err == 0 ? PAL_SUCCESS : SystemNative_ConvertErrorPlatformToPal(errno);
+#else
+            return PAL_ENOPROTOOPT;
+#endif /* HAVE_STRUCT_IP_MREQN */
+        }
+    }
+#endif /* WIN32 */
+
```

Comment:
Makes sense.
It would be even nicer if SetMulticastOption would accept a MulticastInterface because the API to set the interface index would be cleaner. Off course, it needs to work on all platforms.
For now I won't touch SetMulticastOption but I'll move the translation to SetSockOpt and call Interop.Sys.SetIPv4MulticastOption from there.